src/Servers/Kestrel/Core/src/Middleware/HttpConnectionMiddleware.cs

```diff
@@ -33,7 +34,7 @@ public Task OnConnectionAsync(ConnectionContext connectionContext)
                 Protocols = _protocols,
                 ServiceContext = _serviceContext,
                 ConnectionFeatures = connectionContext.Features,
-                MemoryPool = memoryPoolFeature.MemoryPool,
+                MemoryPool = memoryPoolFeature?.MemoryPool ?? System.Buffers.MemoryPool<byte>.Shared,
```

Comment:
We should let this be null (without the null ref that is). Don't fallback to the default memory pool though. If that breaks we should fix those places that use the pool directly.

We also need to add a test for this to make sure it doesn't break.

Reply:
I see what you mean.  So leave it as 
`memoryPool = memoryPoolFeature?`
and solve any side-effects later on in the code path where they occur ?
This could be tested with MemoryEndPoint.

Reply:
Yep.

Reply:
Isn't the  MemoryPoolFeature  accessible in any middleware wired up to Ketsrel  through the HttpConnectionContext  ? If so, any middleware assuming it's presence will break when the endpoint in the instance of Kestrel is changed to a non-socket based endpoint like MemoryEndPoint.

Reply:
Yes but this isn't a required feature so I'd like to understand the places that rely on it being non-null.

Reply:
>this isn't a required feature 

I see, I guess that's the critical bit.  Other middleware shouldn't be making this assumption either.

Reply:
Right.

Reply:
Why not include `?? MemoryPool<byte>.Shared` here? The same thing will happen implicitly later when we new up PipeOptions.

> Yes but this isn't a required feature so I'd like to understand the places that rely on it being non-null.

By adding `?? MemoryPool<byte>.Shared`, the middleware is no longer relying on the feature.

Reply:
What middleware is that? This is a specific middleware (the HTTP middleware) that relies on it currently. The TLS middleware already handles this.

Before doing this fix I just want to understand what code in the HTTP layer relies on this being non-null. I have no problem putting it back.

Reply:
What's the point? Like you say, this is the HTTP middleware. By adding this, there would be no code in the HTTP  layer that relies on this being non-null.

Reply:
>I just want to understand what code in the HTTP layer relies on this being non-null
>I have no problem putting it back

Would I be right in thinking that these two actions are independent and that one is not a blocker for the other ? I guess the first can be achieved by setting the MemoryPool to null and seeing where the Tests fail ( I did this and there were at least 13 failures)
and the second can be done to introduce the fix but without any significant hindrance to carrying out the first (given the change is trivial) ?

A bit of context (no pun intended), the MemoryPool was set in the ConnectionContext when binding to a socket but was not set in the ConnectionContext when binding to a (Bedrock) MemoryEndpoint. In both cases ConnectionContext was intended to pushed down the HttpMiddleware.

Is the Transport factory obliged to set the MemoryPool ? If so, how could it be enforced ? If not then I guess provision of a fallback default is mandatory and why is it set in the ConnectionContext in kestrel ?

Reply:
@halter73 

> What's the point? Like you say, this is the HTTP middleware. By adding this, there would be no code in the HTTP layer that relies on this being non-null.

The transport isn't required to set a pool and the IMemoryPoolFeature isn't required. Kestrel needs to take that into account. Kestrel can also polyfill by defaulting to using a pool but I have 2 concerns:
1. We don't have an understanding where it does that. I'd like to understand before making assumptions.
2. We probably want to use the SlabMemoryPool and not the default memory pool in the places where we explicitly allocate. 

@Ninds

> Would I be right in thinking that these two actions are independent and that one is not a blocker for the other ? I guess the first can be achieved by setting the MemoryPool to null and seeing where the Tests fail ( I did this and there were at least 13 failures)
and the second can be done to introduce the fix but without any significant hindrance to carrying out the first (given the change is trivial) ?

The code change is trivial, measuring the impact isn't without understanding what parts of the code rely on this instance.

> Is the Transport factory obliged to set the MemoryPool ? If so, how could it be enforced ? If not then I guess provision of a fallback default is mandatory and why is it set in the ConnectionContext in kestrel ?

Features can be optional, it's part of the system. It's not set on the ConnectionContext, it's just an optional feature that middleware may use if it exists.